csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs

```diff
@@ -90,8 +103,29 @@ internal static IntPtr Handle
         /// It returns the same instance on every call - `OrtEnv` is singleton
         /// </summary>
         /// <returns>Returns a singleton instance of OrtEnv that represents native OrtEnv object</returns>
-        public static OrtEnv Instance() { return _instance.Value; }
+        public static OrtEnv Instance()
+        {
+            return _instance ?? (_instance = new OrtEnv());
```

Comment:
>return _instance ?? (_instance = new OrtEnv()); [](http://example.com/codeflow?start=12&length=47)

This is not thread-safe. Lazy<> mechanism was used there for good reasons. One of them is thread-safety.

Reply:
I would suggest removing the `Instance()` method fully replacing it with the same structure the java code has:
https://javadoc.io/doc/com.microsoft.onnxruntime/onnxruntime/latest/ai/onnxruntime/OrtEnvironment.html
Only exposing `GetEnvironment` methods. This however may result in some regression.

I can just add a simple `lock` to ensure thread safety.

- [ ] Remove `SafeHandle`?
- [ ] Add `lock` to instantiations
- [ ] Remove `Instance()` method?

What should be done in the scope of this PR?

(Side note, the java version also implements `AutoCloseable` which would have the same possible limitations as `SafeHandle`)

Reply:
Here is how I see the obstacle. We are trying to modify the implementation for the sake of testing in a way that is not acceptable. Here is my suggestion.
Create a `SafeHandle` implementation class that would actually instantiate and destroy the native handle as before. 

The `Env` class would use it to instantiate the native instance and retrieve the handle. It would dispose the implementation instance at the right time. The `Env` class we can simply make `Disposable` instead of `SafeHandle`.

Now, you can have a class that you test directly in your tests for the functionality you want to implement without changing the API or runtime behavior of `Env`.

There are details you will have to iron out. For example, make the class accessible to the test code and some others.

Reply:
It seems as a quite big rewrite of the current `OrtEnv` class, I am a bit hesitant to do this. 

I could add an `internal` class in OrtEnv for the explicit purpose of destroying the singleton instance.
```csharp
internal class OrtEnvInstanceScoper : IDisposable
{
    public OrtEnv Env { get; }

    public OrtEnvInstanceScoper(ThreadingOptions threadingOptions = null)
    {
        Env = threadingOptions == null ? Instance() : GetEnvironment(threadingOptions);
    }
    public void Dispose()
    {
        Env.ReleaseHandle();
        Env.ClearInstance(); // private
    }
}
```